### PR TITLE
Clean up lexer and fix bug

### PR DESCRIPTION
**Pull request by mb64**
This PR cleans up the lexer, making it shorter, more correct, and IMO more readable.

Previously, the lexer failed to lex hexadecimal float literals like `0xabc.def321p0f32` that were type suffixed, but also contain the hex digit sequence `f32` or `f64` beforehand.  This is fixed.

Edit: `happy` does not seem to be very happy in the CI... not sure why

### CODE REVIEW

**Comment by athas** [comment]
Nice catch!

Ignore the `test-unit-and-style` check (it seems to be flaky today), but something is now going wrong with the `tests/hexfloats.fut` test.

File: src/Language/Futhark/Parser/Lexer.x [hunk 1]
```diff
@@ -51,12 +53,12 @@ import Futhark.Util.Loc hiding (L)
...
 @qualunop = (@identifier ".")+ @unop
 
-@opchar = ("+"|"-"|"*"|"/"|"%"|"="|"!"|">"|"<"|"|"|"&"|"^"|".")
-@binop = ("+"|"-"|"*"|"/"|"%"|"="|"!"|">"|"<"|"|"|"&"|"^") @opchar*
+$opchar = [\+\-\*\/\%\=\!\>\<\|\&\^\.]
+@binop = ($opchar # \.) $opchar*
 @qualbinop = (@identifier ".")+ @binop
 
 @space = [\ \t\f\v]
-@doc = "-- |"[^\n]*(\n@space*"--"[^\n]*)*
```
Comment (athas) [inline src/Language/Futhark/Parser/Lexer.x:59]: Can you explain why this change is right?  Won't this match the following?

```
-- | foo bar
baz
```

Or does `.` not match newlines in Alex?

Comment (mb64) [reply src/Language/Futhark/Parser/Lexer.x:59]: Yes – in Alex (like many regexes) `.` is everything but newlines.

[From the docs:](https://www.haskell.org/alex/doc/html/charsets.html)
"The built-in set ‘.’ matches all characters except newline (\n)."

**Comment by athas** [comment]
CI says you have a syntax error in one of the lexer code fragments.

**Comment by mb64** [comment]
Huh, odd.  I thought I had tested it on my machine beforehand, but it was clearly wrong. (`let ... in in ...`)

Should be fixed now

**Comment by athas** [comment]
Thanks for the fix!
